doc/neps/ufunc-overrides.rst

```diff
@@ -153,49 +152,207 @@ tuple in the ``out`` keyword argument.
-  the Ufunc.
+  inputs before outputs, otherwise left to right.
+
+- The first ``__array_ufunc__`` method returning something else than
+  :obj:`NotImplemented` determines the return value of the Ufunc.
 
 - If all ``__array_ufunc__`` methods of the input arguments return
   :obj:`NotImplemented`, a :exc:`TypeError` is raised.
 
-- If a ``__array_ufunc__`` method raises an error, the error is propagated
-  immediately.
-
-If none of the input arguments has an ``__array_ufunc__`` method, the
-execution falls back on the default ufunc behaviour.
+- If a ``__array_ufunc__`` method raises an error, the error is
+  propagated immediately.
+
+- If none of the input arguments had an ``__array_ufunc__`` method, the
+  execution falls back on the default ufunc behaviour.
+
+In the above, there is one proviso: if a class has an
+``__array_ufunc__`` attribute but it is identical to
+``ndarray.__array_ufunc__``, the attribute is ignored.  This happens for
+instances of `ndarray` and for `ndarray` subclasses that did not
+override their inherited ``__array_ufunc__`` implementation.
+
+
+Type casting hierarchy
+----------------------
+
+Similarly to the Python operator dispatch mechanism, writing ufunc
+dispatch methods requires some discipline in order to achieve
+predictable results.
+
+In particular, it is useful to maintain a clear idea of what types can
+be upcast to others, possibly indirectly (i.e. A->B->C is implemented
+but direct A->C not). Moreover, one should make sure the implementations of
+``__array_ufunc__``, which implicitly define the type casting hierarchy,
+don't contradict this.
+
+It is useful to think of the typecasting hierarchy as a graph (see
+example below) in which, for any given class A, all other classes that
+define ``__array_ufunc__`` must belong to exactly one of three groups
+(making this an directed acyclic graph):
+
+- *Above A*: their ``__array_ufunc__`` can handle class A or some
+  member of the "above A" classes. In other words, these are the types
+  that A can be (indirectly) upcast to in ufuncs.
+
+- *Below A*: they can be handled by the ``__array_ufunc__`` of class A
+  or the ``__array_ufunc__`` of some member of the "below A" classes. In
+  other words, these are the types that can be (indirectly) upcast to A
+  in ufuncs.
+
+- *Incompatible*: neither above nor below A; types for which no
+  (indirect) upcasting is possible.
+
+Given this grouping, to ensure that expressions involving ufuncs either
+raise a :exc:`TypeError`, or have a result type that is independent of
+what ufuncs were called, what order they were called in, and what order
+their arguments were in, the above implies that ``__array_ufunc__`` for
+type A should:
+
+- Return an object of type A if all other arguments are of types below A.
+
+- Return :obj:`NotImplemented` if any argument has a type that is above
+  A or with which it is incompatible.
+
+Note that there are, as always, exceptions.  For instance, for a
```

Comment:
I'd love to be clearer here, so maybe we can work out the example better, since for `Quantity`, the logic is quite simple:
```
from astropy import units as u
a = 1.
q = 1.*u.m
type(q + q)
# Quantity
type (q > q)
# ndarray
type(a > (a + q))
# UnitsError: Can only apply 'add' function to dimensionless quantities when other argument is not a quantity (unless the latter is all zero/infinity/nan)
dq = 1.*u.dimensionless_unscaled
type(a > (a + dq))
# ndarray
```
Would it make sense to see `Quantity` as two types, with and without dimension (and there is an exception for all-zero, inf, nan, mostly because this makes many more functions work...)